recirq/quantum_chess/quantum_board.py

```diff
@@ -331,16 +334,81 @@ def _generate_accumulations(self, repetitions: int = 1000) -> None:
 
         self.accumulations_repetitions = repetitions
 
+    def _caching_supported(self, m: move.Move):
+        """Checks if caching is supported for this move."""
+
+        # Caching is supported for a split jump from one full square to two empty squares.
+        if m.move_type == enums.MoveType.SPLIT_JUMP and self.probabilities[
+                square_to_bit(
+                    m.source)] == 1 and self.probabilities[square_to_bit(
+                        m.target)] == 0 and self.probabilities[square_to_bit(
+                            m.target2)] == 0:
+            return True
+        return False
+
+    def cache_results(self, cache_key: CacheKey):
+        if cache_key in self.cache:
+            return
+        if cache_key.move_type == enums.MoveType.SPLIT_JUMP:
+            helper_board = CirqBoard(self.init_basis_state, self.sampler,
+                                     self.device, self.error_mitigation,
+                                     self.noise_mitigation,
+                                     self.transformer if self.device else None)
+            sample_jump_move = move.Move('b1',
+                                         'c3',
+                                         target2='a3',
+                                         move_type=enums.MoveType.SPLIT_JUMP,
+                                         move_variant=enums.MoveVariant.BASIC)
+            helper_board.do_move(sample_jump_move)
+            probs = helper_board.get_probability_distribution(
+                cache_key.repetitions, use_cache=False)
+            self.cache[cache_key] = {
+                "source": 0,
+                "target": probs[square_to_bit(sample_jump_move.target)],
+                "target2": probs[square_to_bit(sample_jump_move.target2)]
+            }
+            self.debug_log += helper_board.debug_log
+
+    @staticmethod
+    def _apply_cache(probability, m, cache_value):
+        for k, v in cache_value.items():
+            square = getattr(m, k)
+            probability[square_to_bit(square)] = v
+        return probability
+
     def get_probability_distribution(self,
-                                     repetitions: int = 1000) -> List[float]:
+                                     repetitions: int = 1000,
+                                     use_cache=True) -> List[float]:
         """Returns the probability of a piece being in each square.
 
         The values are returned as a list in the same ordering as a
         bitboard.
         """
+        move_history_cache_key = (tuple(self.move_history), repetitions)
```

Comment:
As it stands `get_full_squares_bitboard` won't be in sync with the cached results and will trigger a new run of the circuit. Maybe this stuff should be in `_generate_accumulations` so that the full/empty bitboard stuff gets updated too?